src/ows/src/main/java/org/geoserver/ows/Dispatcher.java

```diff
@@ -327,37 +327,49 @@ Request init(Request request) throws ServiceException, IOException {
             } else if (reqContentType != null
                     && ServletFileUpload.isMultipartContent(httpRequest)) {
                 // multipart form upload
-                ServletFileUpload up = new ServletFileUpload();
-                up.setFileItemFactory(new DiskFileItemFactory());
+                ServletFileUpload up = new ServletFileUpload(new DiskFileItemFactory());
 
                 // treat regular form fields as additional kvp parameters
                 Map<String, FileItem> kvpFileItems =
                         new CaseInsensitiveMap<>(new LinkedHashMap<>());
                 try {
-                    for (FileItem item : up.parseRequest(httpRequest)) {
+                    List<FileItem> items = up.parseRequest(httpRequest);
+                    FileItemCleanupCallback.setFileItems(items);
+                    FileItem body = null;
+                    for (FileItem item : items) {
                         if (item.isFormField()) {
-                            kvpFileItems.put(item.getFieldName(), item);
+                            FileItem old = kvpFileItems.put(item.getFieldName(), item);
+                            if (old != null) {
+                                old.delete(); // the temp file can be deleted at this point
+                            }
                         } else {
-                            request.setInput(fileItemReader(item));
+                            if (body != null) {
+                                // GeoServer only reads the last uploaded file per request so any
+                                // other file uploads can be deleted immediately
+                                body.delete();
```

Comment:
The callback is already going to delete all items at the end of the request, so this extra logic improves deletion by a little bit and only for corner cases. Is it worth the extra complication? Makes me wonder if you're trying to address some sort of attack.

Reply:
Deleting temp files that could contain arbitrary contents uploaded by anyone with access to the GeoServer instance seems like a good idea but it's your call on whether or not the extra logic is worth having.  Even with this extra logic, someone with local access to the GeoServer machine could still use GeoServer to upload arbitrary files onto the machine by creating a custom client that sends a request to upload the file and then throttles the request on a subsequent field until the user has time to copy the file out of the temp directory.